tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+            Header::new(":path", &path),
+            Header::new("user-agent", USER_AGENT),
+        ];
+
+        if let Some(body) = &body {
+            hdrs.push(Header::new("content-length", &body.len().to_string()));
+        }
+
+        HttpBinReq {
+            url: url.clone(),
+            hdrs,
+            body,
+            expect_resp_hdrs: expect_resp_hdrs.clone(),
+            resp_hdrs: Vec::new(),
+            resp_body: Vec::new(),
+        }
+    }
+
+    pub fn assert_hdrs(&self) -> bool {
+        let mut ok = true;
+
+        if let Some(expect_hdrs) = &self.expect_resp_hdrs {
+            for hdr in expect_hdrs {
+                if let Some(h) =
+                    self.resp_hdrs.iter().find(|&x| x.name() == hdr.name())
+                {
+                    ok &= hdr.value() == h.value();
+                    continue;
+                }
+
+                ok &= false;
+            }
+        }
+
+        ok
+    }
+}
+
+// A rudimentary structure to hold httpbin response data
+#[derive(Debug, serde::Deserialize)]
+struct HttpBinResponseBody {
+    args: Option<HashMap<String, String>>,
+    data: Option<String>,
+    files: Option<HashMap<String, String>>,
+    form: Option<HashMap<String, String>>,
+    headers: Option<HashMap<String, String>>,
+    json: Option<HashMap<String, String>>,
+    #[serde(rename = "user-agent")]
+    user_agent: Option<String>,
+    server: Option<String>,
+    #[serde(rename = "content-type")]
+    content_type: Option<Vec<String>>,
+    origin: Option<String>,
+    url: Option<String>,
+}
+
+// A helper function pointer type for assertions.
+// Each test takes a vector of HttpBinReq for validation
+// and must return a pass or fail.
+type HttpBinTestAssert = (fn(&[HttpBinReq]) -> bool);
+
+// The main object for getting things done. The factory method
+// new() is used to set up a vector of HttpBinReq objects
+// and map them to a test assertion function. The public functions
+// are used to send requests and store response data. Internally
+// we track some other state to make sure everything goes smoothly.
+//
+// Many tests have similar inputs or assertions, so utility functions
+// help cover many of the common cases like testing different status
+// codes or checking that a response body is echoed back.
+pub struct HttpBinTest {
+    endpoint: url::Url,
+    reqs: Vec<HttpBinReq>,
+    assert: HttpBinTestAssert,
+    issued_reqs: HashMap<u64, usize>,
+    concurrent: bool,
+    current_idx: usize,
+}
+
+impl HttpBinTest {
+    pub fn new(
+        endpoint: url::Url, reqs: Vec<HttpBinReq>, assert: HttpBinTestAssert,
+    ) -> HttpBinTest {
+        HttpBinTest {
+            endpoint,
+            reqs,
+            assert,
+            issued_reqs: HashMap::new(),
+            concurrent: true,
+            current_idx: 0,
+        }
+    }
+
+    /// Set test request concurrency
+    pub fn set_concurrency(&mut self, concurrent: bool) {
```

Comment:
It doesn't look like the concurrency needs to be changed at runtime, so maybe it should just be an argument of `new()`?

Reply:
This is hangover from previous design, so yeah probably can be nuked.